### PR TITLE
Implement lookup event with num_topics

### PR DESCRIPTION
**Pull request by JonoPrest**
Hey @ozgrakkurt,

I'm needing to use the new version of hs client for the indexer and unfortunately there is a regression in handling events with same topic0 but different number of other topics.

I've implemented a version that supports this here. Let me know if the changes are suitable.

### CODE REVIEW

File: hypersync-client/src/decode.rs [hunk 1]
```diff
@@ -1,13 +1,21 @@
...
+use std::collections::HashMap;
 
-use crate::simple_types::Log;
+#[derive(Debug, Hash, Eq, PartialEq)]
+struct EventKey {
+    topic0: Vec<u8>,
+    num_topics: usize,
+}
+
+type DecoderMap = HashMap<EventKey, DynSolEvent>;
```
Comment (JonoPrest) [inline hypersync-client/src/decode.rs:13]: I thought a hashmap would be more performant (at least for larger abis)

File: hypersync-client/src/decode.rs [hunk 2]
```diff
@@ -18,33 +26,20 @@ impl Decoder {
...
-
-        map.dedup_by(|a, b| a.0 == b.0);
-
-        let new_len = map.len();
-
-        if initial_len != new_len {
-            return Err(anyhow!(
-                "duplicate event signature selectors (topic0) found.
-This might be because the 'indexed' keyword doesn't effect the selector of an event signature."
-            ));
```
Comment (JonoPrest) [inline hypersync-client/src/decode.rs:45]: Removes the need for this code. Hashmap will deduplicate and account for different indexed keywords with same topic 0

File: hypersync-client/src/decode.rs [hunk 3]
```diff
@@ -69,8 +64,22 @@ This might be because the 'indexed' keyword doesn't effect the selector of an ev
...
+            num_topics: topics.iter().fold(
+                0,
+                |accum, topic| {
+                    if topic.is_some() {
+                        accum + 1
+                    } else {
+                        accum
+                    }
+                },
+            ),
```
Comment (JonoPrest) [inline hypersync-client/src/decode.rs:78]: This fold is the only perf regression I see, but I imagine should be minimal.

File: hypersync-client/src/decode.rs (hunk 1 above)
Comment (ozgrakkurt) [reply hypersync-client/src/decode.rs:13]: Vec will have better performance up to a very big number of elements (500, 1000)

**Review by ozgrakkurt — APPROVED** [review]

**Pull request merged**